### PR TITLE
PWX-37370 Automation deleting VolumeSnapshots and ensuring they are deleted successfully

### PR DESCRIPTION
**Pull request by sayalasomayajula-px**
**What this PR does / why we need it**:
Deploys fada apps, create csi snapshots , restores them and deletes the snapshots and check if those are deleted successfully
**Which issue(s) this PR fixes** (optional)
Closes #PWX-37370

**Special notes for your reviewer**:

PASS LOGS: https://jenkins.pwx.dev.purestorage.com/blue/organizations/jenkins/Users%2Fkrishna%2Ftp-pxe-byoc/detail/tp-pxe-byoc/3177/pipeline
 https://jenkins.pwx.dev.purestorage.com/blue/organizations/jenkins/Users%2Fkrishna%2Ftp-pxe-byoc/detail/tp-pxe-byoc/3174/pipeline/

### CODE REVIEW

File: drivers/scheduler/k8s/k8s.go [hunk 1]
```diff
@@ -4757,6 +4756,36 @@ func (k *K8s) GetSnapshotsInNameSpace(ctx *scheduler.Context, snapshotNameSpace
 	return snapshotList, nil
 }
 
+// IsCsiSnapshotExists checks if snapshot exists in namespace
+func (k *K8s) IsCsiSnapshotExists(ctx *scheduler.Context, snapshotName string, namespace string) (bool, error) {
+	snaplist, err := k.GetSnapshotsInNameSpace(ctx, namespace)
+	log.FailOnError(err, "Failed to get Snapshots in namespace [%v]", namespace)
```
Comment (lsrinivas-pure) [inline drivers/scheduler/k8s/k8s.go:4762]: return the error.. do the validation in the test

File: drivers/scheduler/k8s/k8s.go [hunk 2]
```diff
@@ -4757,6 +4756,36 @@ func (k *K8s) GetSnapshotsInNameSpace(ctx *scheduler.Context, snapshotNameSpace
...
+		}
+	}
+	return false, nil
+
+}
+
+// DeleteCsiSnapshotsFromNamespace delete the snapshots from the namespace
+func (k *K8s) DeleteCsiSnapshotsFromNamespace(ctx *scheduler.Context, namespace string) error {
+	snaplist, err := k.GetSnapshotsInNameSpace(ctx, namespace)
+	log.FailOnError(err, "Failed to get Snapshots in namespace [%v]", namespace)
```
Comment (lsrinivas-pure) [inline drivers/scheduler/k8s/k8s.go:4780]: same here

File: drivers/scheduler/k8s/k8s.go [hunk 3]
```diff
@@ -4757,6 +4756,36 @@ func (k *K8s) GetSnapshotsInNameSpace(ctx *scheduler.Context, snapshotNameSpace
...
+	for _, snap := range snaplist.Items {
+		err = k.DeleteCsiSnapshot(ctx, snap.ObjectMeta.Name, namespace)
+		log.FailOnError(err, "Failed to delete snapshot [%v] in namespace [%v]", snap.ObjectMeta.Name, namespace)
+	}
```
Comment (lsrinivas-pure) [inline drivers/scheduler/k8s/k8s.go:4784]: same here

**Review by Pure-AdamuKaapan — CHANGES_REQUESTED** [review]
All the existing code looks good - have a request to add it to another existing test though.

File: drivers/scheduler/dcos/dcos.go [hunk 4]
```diff
@@ -134,14 +134,32 @@ func (d *dcos) DeleteSnapShot(ctx *scheduler.Context, snapshotName, snapshotName
 }
 
 // GetSnapshotsInNameSpace get the snapshots list for the namespace
-func (d *dcos) GetSnapshotsInNameSpace(ctx *scheduler.Context, snapshotNameSpace string) (*snapv1.VolumeSnapshotList, error) {
+func (d *dcos) GetSnapshotsInNameSpace(ctx *scheduler.Context, snapshotNameSpace string) (*volsnapv1.VolumeSnapshotList, error) {
```
Comment (kphalgun-px) [inline drivers/scheduler/dcos/dcos.go:137]: I think it should be `*snapv1.VolumeSnapshotList` and not `*volsnapv1.VolumeSnapshotList`. Please verify.

File: drivers/scheduler/k8s/k8s.go [hunk 5]
```diff
@@ -4757,6 +4756,45 @@ func (k *K8s) GetSnapshotsInNameSpace(ctx *scheduler.Context, snapshotNameSpace
...
+	return false, nil
+
+}
+
+// DeleteCsiSnapshotsFromNamespace delete the snapshots from the namespace
+func (k *K8s) DeleteCsiSnapshotsFromNamespace(ctx *scheduler.Context, namespace string) error {
+	snaplist, err := k.GetSnapshotsInNameSpace(ctx, namespace)
+	if err != nil {
+		log.InfoD("Failed to get Snapshots in namespace [%v]", namespace)
```
Comment (kphalgun-px) [inline drivers/scheduler/k8s/k8s.go:4784]: Can we include `ctx.App.Key` in the InfoD message?

File: drivers/scheduler/k8s/k8s.go [hunk 6]
```diff
@@ -4757,6 +4756,45 @@ func (k *K8s) GetSnapshotsInNameSpace(ctx *scheduler.Context, snapshotNameSpace
 	return snapshotList, nil
 }
 
+// IsCsiSnapshotExists checks if snapshot exists in namespace
+func (k *K8s) IsCsiSnapshotExists(ctx *scheduler.Context, snapshotName string, namespace string) (bool, error) {
+	snaplist, err := k.GetSnapshotsInNameSpace(ctx, namespace)
+	if err != nil {
+		log.InfoD("Failed to get Snapshots in namespace [%v]", namespace)
```
Comment (kphalgun-px) [inline drivers/scheduler/k8s/k8s.go:4763]: Can we include `ctx.App.Key` in the InfoD message?

File: tests/basic/pure_test.go [hunk 7]
```diff
@@ -4340,3 +4341,92 @@ var _ = Describe("{CreateCloneOfTheFADAVolume}", func() {
...
+		StartTorpedoTest("CreateCsiSnapshotsforFADAandDelete",
+			"Create CSI snapshots for FADA volumes and delete them", nil, 0)
+	})
+	itLog := "CreateCsiSnapshotsforFADAandDelete"
+	It(itLog, func() {
+		log.InfoD(itLog)
+		var volSnapshotClass *volsnapv1.VolumeSnapshotClass
+		var volumeSnapshotMap map[string]*volsnapv1.VolumeSnapshot
+		applist := Inst().AppList
+		defer DestroyApps(contexts, nil)
```
Comment (kphalgun-px) [inline tests/basic/pure_test.go:4368]: Can we defer DestroyApps after contexts is initialised?

File: tests/basic/pure_test.go [hunk 8]
```diff
@@ -4340,3 +4341,92 @@ var _ = Describe("{CreateCloneOfTheFADAVolume}", func() {
...
+		Step(stepLog, func() {
+			log.InfoD(stepLog)
+			snapShotClassName := PureSnapShotClass
+			volSnapshotClass, err = Inst().S.CreateCsiSnapshotClass(snapShotClassName, "Delete")
+			log.FailOnError(err, "Failed to create volume snapshot class")
+			log.InfoD("Successfully created volume snapshot class: %v", volSnapshotClass.Name)
+		})
+		stepLog = "Creating snapshots for all apps in the context and validate them"
+		Step(stepLog, func() {
+			for _, ctx := range contexts {
```
Comment (kphalgun-px) [inline tests/basic/pure_test.go:4396]: log.InfoD is missing

File: tests/basic/pure_test.go [hunk 9]
```diff
@@ -4340,3 +4341,92 @@ var _ = Describe("{CreateCloneOfTheFADAVolume}", func() {
...
+				log.FailOnError(err, "Failed to validate the snapshots")
+			}
+		})
+		stepLog = "Delete the snapshots created for all apps in the context"
+		Step(stepLog, func() {
+			log.InfoD(stepLog)
+			for _, ctx := range contexts {
+				err := Inst().S.DeleteCsiSnapshotsFromNamespace(ctx, ctx.App.NameSpace)
+				log.FailOnError(err, "Failed to delete the snapshots")
+				log.InfoD("Deleted the snapshots successfully")
```
Comment (kphalgun-px) [inline tests/basic/pure_test.go:4409]: I think 

```go
log.InfoD("Deleted the snapshots successfully")
```

should be outside the for loop.

File: tests/basic/pure_test.go [hunk 10]
```diff
@@ -4340,3 +4341,92 @@ var _ = Describe("{CreateCloneOfTheFADAVolume}", func() {
...
+			}
+		})
+		stepLog = "Validate all snapshots are deleted "
```
Comment (kphalgun-px) [inline tests/basic/pure_test.go:4412]: Should we wait after deleting the snapshot before validating?

File: drivers/scheduler/scheduler.go [hunk 11]
```diff
@@ -225,7 +225,13 @@ type Driver interface {
...
 
 	// GetSnapshotsInNameSpace get the snapshots list for the namespace
-	GetSnapshotsInNameSpace(ctx *Context, snapshotNameSpace string) (*snapv1.VolumeSnapshotList, error)
+	GetSnapshotsInNameSpace(ctx *Context, snapshotNameSpace string) (*volsnapv1.VolumeSnapshotList, error)
+
+	//DeleteCsiSnapshotsFromNamespace deletes the all snapshots from a namespace
+	DeleteCsiSnapshotsFromNamespace(ctx *Context, namespace string) error
+
+	//IsCsiSnapshotExists checks if a snapshot exists in the particular namespace
+	IsCsiSnapshotExists(ctx *Context, snapshotName string, namespace string) (bool, error)
```
Comment (kphalgun-px) [inline drivers/scheduler/scheduler.go:234]: We can get namespace using `ctx.App.NameSpace`. How about we remove namespace argument to reduce redundancy?

Comment (kphalgun-px) [reply drivers/scheduler/scheduler.go:234]: May be we can ignore this for now as it is consistent with GetSnapshotsInNameSpace. We can take this up in future.

Comment (sayalasomayajula-px) [reply drivers/scheduler/scheduler.go:234]: we need this to validate each individual snapshots

File: tests/basic/pure_test.go [hunk 12]
```diff
@@ -4340,3 +4341,92 @@ var _ = Describe("{CreateCloneOfTheFADAVolume}", func() {
 		appsValidateAndDestroy(contexts)
 	})
 })
+
+var _ = Describe("{CreateCsiSnapshotsforFADAandDelete}", func() {
```
[the remark this answers is not in the document]
Comment (sayalasomayajula-px) [reply tests/basic/pure_test.go:4345]: added

**Review by lsrinivas-pure — APPROVED** [review]
LGTM

**Review by Pure-AdamuKaapan — APPROVED** [review]
LGTM!

**Pull request merged**